common/src/main/java/com/graphicsfuzz/common/util/FragmentShaderJobToVertexShaderJob.java

```diff
@@ -124,44 +149,55 @@ void main() {
-                    new BinaryExpr(
-                    new ParenExpr(
-                    new BinaryExpr(
-                        new VariableIdentifierExpr(Constants.GLF_POS),
-                        new TypeConstructorExpr(BasicType.VEC4.toString(),
-                            new FloatConstantExpr("1.0"),
-                            new FloatConstantExpr("1.0"),
-                            new FloatConstantExpr("0.0"),
-                            new FloatConstantExpr("0.0")),
-                        BinOp.ADD)),
-                        new TypeConstructorExpr(BasicType.VEC4.toString(),
-                            new FloatConstantExpr("128.0"),
-                            new FloatConstantExpr("128.0"),
-                            new FloatConstantExpr("1.0"),
-                            new FloatConstantExpr("1.0")),
-                        BinOp.MUL),
-            BinOp.ASSIGN)));
+    if (analysis.usesFragCoord) {
+      // Add a global _GLF_FragCoord that will replace gl_FragCoord
+      vertexShader.addDeclarationBefore(
+          new VariablesDeclaration(
+              BasicType.VEC4, new VariableDeclInfo(Constants.GLF_FRAGCOORD, null, null)),
+          firstNonPrecisionDeclaration);
+
+      // Add vertex position input
+      vertexShader.addDeclarationBefore(
+          new VariablesDeclaration(new QualifiedType(BasicType.VEC4,
+              Arrays.asList(new LayoutQualifierSequence(
+                      new LocationLayoutQualifier(0)),
+                  TypeQualifier.SHADER_INPUT)),
+              new VariableDeclInfo(Constants.GLF_POS, null, null)),
+          firstNonPrecisionDeclaration);
 
+      // At start of main, populate _GLF_FragCoord with coordinates calculated from vertex input
+      vertexShader.getMainFunction().getBody().insertStmt(0,
+          new ExprStmt(
+              new BinaryExpr(
+                  new VariableIdentifierExpr(Constants.GLF_FRAGCOORD),
+                  new BinaryExpr(
+                      new ParenExpr(
+                          new BinaryExpr(
+                              new VariableIdentifierExpr(Constants.GLF_POS),
+                              new TypeConstructorExpr(BasicType.VEC4.toString(),
+                                  new FloatConstantExpr("1.0"),
+                                  new FloatConstantExpr("1.0"),
+                                  new FloatConstantExpr("0.0"),
+                                  new FloatConstantExpr("0.0")),
+                              BinOp.ADD)),
+                      new TypeConstructorExpr(BasicType.VEC4.toString(),
+                          new FloatConstantExpr("128.0"),
+                          new FloatConstantExpr("128.0"),
+                          new FloatConstantExpr("1.0"),
+                          new FloatConstantExpr("1.0")),
+                      BinOp.MUL),
+                  BinOp.ASSIGN)));
+    }
+
+    final IParentMap parentMap = IParentMap.createParentMap(vertexShader);
```

Comment:
Please add a comment to explain the overall purpose of the anonymous visitor that follows.

Reply:
done